File: rust/rbw-core/src/rabbitmq.rs
use std::time::Duration;

use lapin::{
    options::{
        BasicPublishOptions, QueueDeclareOptions,
    },
    types::FieldTable,
    BasicProperties, Channel, Connection, ConnectionProperties,
};
use tokio::time::sleep;
use tracing::{info, warn};

use crate::config::{rabbitmq_host, DynError, JOBS_QUEUE, RESULTS_QUEUE};

pub async fn connect_with_retry(retries: u32, delay_secs: u64) -> Result<Connection, DynError> {
    let host = rabbitmq_host();
    let addr = format!("amqp://guest:guest@{}:5672/%2f", host);

    let mut last_error: Option<String> = None;

    for attempt in 1..=retries {
        match Connection::connect(&addr, ConnectionProperties::default()).await {
            Ok(conn) => return Ok(conn),
            Err(err) => {
                last_error = Some(err.to_string());
                warn!(
                    "[connect] attempt {}/{} failed: {}",
                    attempt, retries, err
                );
                sleep(Duration::from_secs(delay_secs)).await;
            }
        }
    }

    Err(format!(
        "Could not connect to RabbitMQ after {} attempts: {}",
        retries,
        last_error.unwrap_or_else(|| "unknown error".to_string())
    )
    .into())
}

pub async fn declare_queues(channel: &Channel) -> Result<(), DynError> {
    channel
        .queue_declare(
            JOBS_QUEUE,
            QueueDeclareOptions {
                durable: true,
                ..Default::default()
            },
            FieldTable::default(),
        )
        .await?;

    channel
        .queue_declare(
            RESULTS_QUEUE,
            QueueDeclareOptions {
                durable: true,
                ..Default::default()
            },
            FieldTable::default(),
        )
        .await?;

    Ok(())
}

pub async fn publish_bytes(channel: &Channel, queue_name: &str, payload: &[u8]) -> Result<(), DynError> {
    channel
        .basic_publish(
            "",
            queue_name,
            BasicPublishOptions::default(),
            payload,
            BasicProperties::default()
                .with_delivery_mode(2)
                .with_content_type("application/octet-stream".into()),
        )
        .await?
        .await?;

    Ok(())
}

pub async fn create_channel() -> Result<Channel, DynError> {
    let conn = connect_with_retry(30, 2).await?;
    let channel = conn.create_channel().await?;
    declare_queues(&channel).await?;
    info!("Connected to RabbitMQ");
    Ok(channel)
}

File: python/src/bw/messaging/rabbitmq.py
import logging
import os
import socket
import time

import pika


RABBITMQ_HOST = os.getenv("RABBITMQ_HOST", "rabbitmq")
JOBS_QUEUE = "jobs"
RESULTS_QUEUE = "results"


def get_hostname() -> str:
    return socket.gethostname()


def connect_with_retry(
    retries: int = 30,
    delay: float = 2.0,
    logger: logging.Logger | None = None,
) -> pika.BlockingConnection:
    last_error = None

    for attempt in range(1, retries + 1):
        try:
            parameters = pika.ConnectionParameters(
                host=RABBITMQ_HOST,
                heartbeat=60,
            )
            connection = pika.BlockingConnection(parameters)
            return connection
        except Exception as exc:
            last_error = exc
            message = f"[connect] attempt {attempt}/{retries} failed: {exc}"
            if logger is not None:
                logger.warning(message)
            else:
                print(message)
            time.sleep(delay)

    raise RuntimeError(
        f"Could not connect to RabbitMQ after {retries} attempts: {last_error}"
    )


def declare_queues(channel: pika.adapters.blocking_connection.BlockingChannel) -> None:
    channel.queue_declare(queue=JOBS_QUEUE, durable=True)
    channel.queue_declare(queue=RESULTS_QUEUE, durable=True)


def publish_bytes(channel, queue_name: str, payload: bytes) -> None:
    channel.basic_publish(
        exchange="",
        routing_key=queue_name,
        body=payload,
        properties=pika.BasicProperties(
            delivery_mode=2,
            content_type="application/octet-stream",
        ),
    )